### ISSUE TITLE
Add ability to create unique use-once survey links 

### ISSUE THREAD

**Issue by penyuan** [snoopForms/snoopforms#72]
### Is your proposal related to a problem?

There is a need to create surveys where individual people can only submit a single response without employing invasive tracking.

### Describe the solution you'd like

Once a survey is in place, allow the creation of an arbitrary number of unique, use-only-ones links to the survey. I can then distribute these links to respondents, where they can use the unique links to respond. Once a link has been used, it cannot be used again to submit a response.

### Describe alternatives you've considered

There are ways to track responses by users through fingerprinting or tracking IPs, and other things. All of these invade privacy, and should be solved through these unique links instead.

**Comment by ty-kerr** [comment]
id like to work on this!

**Comment by mattinannt** [comment]
This is how I would propose the implementation:

This is a feature for Link-Surveys only.
A unique id can be added to a survey link via search parameters, e.g. `?id=abcd12`.
The response schema in the database (`schema.prisma`) needs to be extended with a new field `externalId String?` as well as a constraint `@@unique([surveyId, externalId])`.

Please make sure that the Link Survey is only displayed when no `id` is set in search parameters or if there isn't already a response for this `id`. If a response for this `id` already exists, please show a message to the user indicating that she/he has already given a response to this survey.

**Comment by ty-kerr** [comment]
ill pick this one up next!

**Comment by ty-kerr** [comment]
@mattinannt would it be a good idea to allow users to select if they want to share a unique link when they open the share survey link dialog? or should we default all survey links to be unique

**Comment by ykburda** [comment]
Hi,
If we don’t select individual survey in the UI how will we seperate anonymous entry in that survey? So I think it should be an option.

**Comment by ykburda** [comment]
Also we discussed it would be something like `?userId=abc&responseId=123` and we assign the response to a person and also give it a unique id to make sure it is only answered once.

If you allow “userid” in the link this is what we want to see in responsed by.
If you allow “responseid” in the link we would like to see this id in the results in the screen of formbricks too

**Comment by ykburda** [comment]
It would be better to put an option in survey setting "Response options" tab whether it is individual survey or anonymous survey.

**Comment by jobenjada** [comment]
We'll discuss and follow up. Are you commited to build this and contribute it back? :)

**Comment by ykburda** [comment]
@ty-kerr is assigned to build the functionality. But anytime i can contribute on requirement.

**Comment by jobenjada** [comment]
@ykburda tykerr is busy with his job so he wont be able to complete this feature. If you'd like to pick it up, that'd be appreciated :)

Here is my current understanding of the need for this feature:

**Problems**
- Allow the creation of single-use survey links
- Allow identification of respondents without tracking

**Current possibilities with Formbricks**
- [Identify users by ID in link survey:](https://formbricks.com/docs/link-surveys/user-identification) We've built this so that you can attach a link survey response to a user in your DB or create a new user profile with the ID in the link.

This allows you to identify a user without having to track anything. You can set a custom ID which will then create a new user in the People view. The individual Response will not be marked as "Anonymous" anymore but link to the user profile.

**Solving single-use links**
Add another option to the Response Options for Link Surveys only:

<img width="1012" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/af138a3b-78c2-45f3-b5ec-bffd15c2db83">

How it works:
- If the link URL does not contain a single use Id, the "Survey Link Used" message will be shown
- If a submission with the single use Id from the URL already exists, the same message will be shown

Example: https://app.formbricks.com/s/ca2da...asd**?suId=SingleUseId**

With user identificiation: https://app.formbricks.com/s/ca2da...asd**?suId=SingleUseId&userId=user123**

**Out of scope:**
- Compare the suId with a predefined list of unique Ids to only allow responses from these Ids. This means that if a user is clever enough to manually edit the suId in the URL to submit another response, thats possible.

**Summary Page**
Add the single use id in the top right next to the 'completed' tag and the date:
<img width="1183" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/3831615a-a6f8-4e91-84aa-d5ec21fdc90f">
FYI the "Anonymous" gets replaced by the userId when its set


**How to build this**
- Add a 'suId' string field to the model Response to store the custom single use id. Add a senseful comment to the field.
- Add a 'singleUse' Object with a singleUseEnabled boolean, header and subheader as strings to the model Survey.
- When adding the logic in the LinkSurvey.tsx make sure to only run add it when the boolean is true to prevent slowing down every link survey. Keep it performant :)


**Testing**
We have:
1. Identification via URL parameter
2. Link prefilling of responses via URL parameter
3. Single Use Ids

Test if all of the above work well together.

@ykburda Is this what you have in mind? Can your team pick this up and contribute it back to the main repo?
@penyuan Your case is already possible :)

**Comment by ykburda** [comment]
Hi,
You have described all details very well. Only I prefer different messages on cases below.
When option enabled;
- If the link URL does not contain a single use Id. It can be “This survey can only be taken by invitation.”
- If a submission with the single use Id from the URL already exists. It can be "The survey has already been answered".

And the last thing, I'm not sure i can complete all these in coding myself.Sorry I can't pick it up for development.
